lib/debuggerTypes.ml

```diff
@@ -0,0 +1,54 @@
+(************************************************************************)
+(*         *   The Coq Proof Assistant / The Coq Development Team       *)
+(*  v      *         Copyright INRIA, CNRS and contributors             *)
+(* <O___,, * (see version control and CREDITS file for authors & dates) *)
+(*   \VV/  **************************************************************)
+(*    //   *    This file is distributed under the terms of the         *)
+(*         *     GNU Lesser General Public License Version 2.1          *)
+(*         *     (see LICENSE file for the text of the license)         *)
+(************************************************************************)
+
+(** Debugger-defined types useful for communicating with IDEs *)
+
+(** The type of coqtop goals *)
+type goal = {
+  goal_id : string;
+  (** Unique goal identifier *)
+  goal_hyp : Pp.t list;
+  (** List of hypotheses *)
+  goal_ccl : Pp.t;
+  (** Goal conclusion *)
+  goal_name : string option;
+  (** User-level goal name *)
+}
+
+(** Subset of goals to print. *)
+type goal_flags = {
+   gf_mode : string;
+   gf_fg : bool;
+   gf_bg : bool;
+   gf_shelved : bool;
+   gf_given_up : bool;
+ }
+
+type 'a pre_goals = {
+  fg_goals : 'a list;
+  (** List of the focused goals *)
+  bg_goals : ('a list * 'a list) list;
+  (** Zipper representing the unfocused background goals *)
+  shelved_goals : 'a list;
+  (** List of the goals on the shelf. *)
+  given_up_goals : 'a list;
+  (** List of the goals that have been given up *)
+}
+
+type goals = goal pre_goals
+
+(* specific return types *)
+
+type subgoals_rty = goals option
+type goals_rty = goals option
+type db_stack_rty = (string * (string * int list) option) list
+type db_vars_rty = (string * Pp.t) list
+
+let in_debug = ref false (* tells whether we're in the Ltac debugger or not *)
```

Comment:
Please don't use a global reference for this.

Reply:
It's the only reasonable way to deal with the differences between the xml protocol and what you've insisted on for DebugHook.  The xml protocol acts as an RPC--each incoming message is handled by a function whose return value is the reply message.  Unlike DebugHook.

Globals are a standard OCaml construct.  Coq could not exist without them.

I haven't worked on the debugger since March.  At this point it's tempting to abandon further work on it because the review process is so bad--it takes far too long and there are too many protracted debates about unimportant details.  It  (IMO unnecessarily) takes so much calendar time that it's very hard to stay motivated.  I think you'd like to see more improvements to the debugger.  There's a tradeoff between insisting on every detail versus having more developers working to improve Coq.

Let me know your thoughts.

Reply:
Indeed there is a trade-off between using more maintenable patterns or using hacky approaches. The work on the debugger is appreciated but we need to ensure it doesn't increase the technical debt and maintenance burden for the future.

Having a pure message-passing protocol depend on a random reference is a sure recipe for problems, and indeed we have had this discussion many times.

Why do you need the reference on the first place? The XML protocol should just reflect the DebugHook protocol, what are the differences you need to take care of?

Reply:
> I haven't worked on the debugger since March. At this point it's tempting to abandon further work on it because the review process is so bad--it takes far too long and there are too many protracted debates about unimportant details. It (IMO unnecessarily) takes so much calendar time that it's very hard to stay motivated. I think you'd like to see more improvements to the debugger. There's a tradeoff between insisting on every detail versus having more developers working to improve Coq.
> 
> Let me know your thoughts.

(sorry to step in, I don't even remember why I'm following this PR and this is probably slightly off topic)

It's indeed a known fact that we are pretty bad at providing timely and efficient reviews. Maybe in part because new PRs easily touch rather technical points making the pool of qualified potential reviewers quite small. Reviewers which can already be busy with other tasks. It is also true that reviewers may insist on a high standard of quality for the new code, particularly regarding functional programming paradigms. This may feel strange as not all the codebase meets those standards but all the reviewers want is for the new code to be as easy to maintain as possible in the future (maybe they should sometimes make more explicit why they think a functional style can make this easier than an imperative one, even when it's perfectly obvious to them).

About the Ltac debugger, maybe the somewhat limited interest is increased by the following factors:
* ltac is pervasive in Coq developments but there are now a few newer/better alternatives to it and many people try to switch to those (for instance, I myself rarely use ltac nowadays, preferring ssreflect for interactive proofs and ltac2 or elpi for automation)
* I may be wrong bu I'm under the impression that you work is tightly linked to Coqide which again is there, useful and used by a large minority of Coq users but is no longer actively developed and some developers even think it should as well die as Coq developers shouldn't spend time maintaining an IDE.

I'm not saying you should target another tactic language or IDE but maybe advertising the fact that your work would be flexible enough to be easy to adapt for instance to ltac2 or vscoq could help atract more interest (and eventually more/better reviews).

Anyway, I'm a bit sad of the current situation. I certainly value your work here, particularly with the reference manual (your docgram tool for instance is instrumental in keeping the manual in sync with the actual syntax of Coq) and I would be deeply saddened if this bad experience with the debugger deters you from further participating in Coq's development.

Summoning @Zimmi48 who is our software engineering expert and may have a more acute point of view on all this.